storage/blockchain/src/ops/tx.rs

```diff
@@ -106,70 +106,69 @@ pub fn add_tx(
     //------------------------------------------------------ Outputs
     // Output bit flags.
     // Set to a non-zero bit value if the unlock time is non-zero.
-    let output_flags = match tx.prefix.timelock {
+    let output_flags = match tx.prefix().additional_timelock {
         Timelock::None => OutputFlags::empty(),
         Timelock::Block(_) | Timelock::Time(_) => OutputFlags::NON_ZERO_UNLOCK_TIME,
     };
 
-    let mut amount_indices = Vec::with_capacity(tx.prefix.outputs.len());
-
-    for (i, output) in tx.prefix.outputs.iter().enumerate() {
-        let key = *output.key.as_bytes();
-
-        // Outputs with clear amounts.
-        let amount_index = if let Some(amount) = output.amount {
-            // RingCT (v2 transaction) miner outputs.
-            if miner_tx && tx.prefix.version == 2 {
+    let amount_indices = match &tx {
+        Transaction::V1 { prefix, .. } => prefix
+            .outputs
+            .iter()
+            .map(|output| {
+                // Pre-RingCT outputs.
+                Ok(add_output(
+                    output.amount.unwrap_or(0),
+                    &Output {
+                        key: output.key.0,
+                        height,
+                        output_flags,
+                        tx_idx: tx_id,
+                    },
+                    tables,
+                )?
+                .amount_index)
+            })
+            .collect::<Result<Vec<_>, RuntimeError>>()?,
+        Transaction::V2 { prefix, proofs } => prefix
+            .outputs
+            .iter()
+            .enumerate()
+            .map(|(i, output)| {
                 // Create commitment.
                 // <https://github.com/Cuprate/cuprate/pull/102#discussion_r1559489302>
                 // FIXME: implement lookup table for common values:
                 // <https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/ringct/rctOps.cpp#L322>
-                let commitment = (ED25519_BASEPOINT_POINT
-                    + monero_serai::H() * Scalar::from(amount))
-                .compress()
-                .to_bytes();
-
+                let commitment = if miner_tx {
+                    ED25519_BASEPOINT_POINT
+                        + *monero_serai::generators::H * Scalar::from(output.amount.unwrap_or(0))
+                } else {
+                    let commitment = proofs
+                        .as_ref()
+                        .expect("A V2 transaction with no RCT proofs is a miner tx")
+                        .base
+                        .commitments[i];
+
+                    commitment
+                };
+
+                (commitment, output.key.0)
+            })
+            .map(|(commitment, key)| {
+                // Add the RCT output.
                 add_rct_output(
                     &RctOutput {
                         key,
```

Comment:
Is there a reason for the double `.map()` here instead of doing everything in 1?
```suggestion
                // Add the RCT output.
                add_rct_output(
                    &RctOutput {
                        key: output.key.0,
```

Reply:
The first map had a early return originally, looks like I removed that and left this.